Document the linting configuration (#821)
elm-tooling/elm-language-server: docs/linting.md

# Linting Configuration

Elm Language Server has built-in support for linting.

> All linting rules are enabled by default. If you want to disable rule you 

Reviewer: ```suggestion
All linting rules are enabled by default. If you want to disable a rule you 
can disable them by placing a configuration file in your project.
```

> All linting rules are enabled by default. If you want to disable rule you 
> an disable them by placing a configuration file in your project.

Reviewer: ```suggestion
All linting rules are enabled by default. If you want to disable a rule you 
can disable them by placing a configuration file in your project.
```

    "UnnecessaryListConcat": true,
    "UnnecessaryPortModule": true,
    "UnusedImport": true,
    "UnusedImportAlias": true,
    "UnusedImportedVariable": true,
    "UnusedPatternVariable": true,
    "UnusedTopLevel": true,
    "UnusedTypeAlias": true,
    "UnusedValueConstructor": true,
    "UnusedVariable": true,
    "UseConsOverConcat": true,
    "MissingTypeAnnotation": true
  },
  "excludedPaths": []
}
```

## Exclude paths

You can exclude certain files and folders from being linted by adding their path to
`excludedPaths` in `elm-analyse.json`.

### Example:

```json
"excludedPaths" : [
	"src/Vendor",
	"src/App/FileThatShouldNotBeLinted.elm"
],
```

## Checks

### `BooleanCase`

If you case over a boolean value, it is more idiomatic to use an if-expression.

#### Example rule violation:

```elm
thing : Boolean -> String
thing x =
    case x of
        True ->
            "Hello"
        False ->
            "Goodbye"
```

### `DropConcatOfLists`

If you concatenate two list literals `[...] ++ [...]`, then you can merge them into one list.

#### Example rule violation:

```elm
foo : List Int
foo =
    [ 1, 2, 3 ] ++ [ 4, 5, 6]
```

### `DropConsOfItemAndList`

> f you cons an item to a literal list (x :: [1, 2, 3]), then you can just put the item into the list.

Reviewer: ```suggestion
If you cons an item to a literal list (x :: [1, 2, 3]), then you can just put the item into the list.
```

Value constructors which are not exposed and used should be eliminated.

#### Example rule violation:

```elm
module Greet exposing (Color, red, blue)

type Color
    = Blue
    | Red
    | Green

red : Color
red = Red

blue : Color
blue = Blue
```

### `UnusedVariable`

Variables that are not used could be removed or marked as '_'.

#### Example rule violation:

```elm
module Foo exposing (foo)

foo : String -> Int
foo x =
    1
```

### `UseConsOverConcat`

If you concatenate two lists, but the right hand side is a single element list, then you should use the cons operator `::`.

#### Example rule violation:

```elm
foo : List String
foo =
    [ "a" ] ++ [ "b", "c" ]
```

### `MissingTypeAnnotation`

Type annotations are required on top-level declarations.

#### Example rule violation:

```elm
module Foo exposing (foo)

foo = 1
```

_Note: This check was called "NoTopLevelSignature" in Elm-Analyse._

## Compatibility with Elm Analyse

The linter was inspired by the tool [Elm-Analyse](https://stil4m.github.io/elm-analyse/) by [Mats Stijlaart](https://github.com/stil4m), and existing elm-analyse configuration should be fully compatible with ElmLS.

> While Elm Language Server does not use Elm-Analyse direct, the check names and descriptions are inpired from it.

Reviewer: ```suggestion
While Elm Language Server does not use Elm-Analyse direct, the check names and descriptions are inspired by it.
```
